Fix D3D12 pixel history MRT shader out results.

jristic opened the pull request:
* Previously the color target was always bound as RT0 regardless of which RT we wanted history for, meaning shader out and post mod results were showing as the values written to RT0.
* The color target is now bound to the slot of the target of history for both shader out and post mod paths, but remains bound to RT0 for the primitive ID path as that shader always writes to RT0.

<!--
Before submitting a pull request you are strongly recommended to read the
docs/CONTRIBUTING.md file which gives some information on how to prepare a
change:

https://github.com/baldurk/renderdoc/blob/v1.x/docs/CONTRIBUTING.md

For small changes you don't have to read the document end to end, but should at
least look at the sections on how to ensure your code and commits are formatted
according to the style requirements.
-->

## Description

<!--
Describe here what your pull request changes and why it should happen. For small
changes which are obvious this can just be a line or two - even the commit
message is sometimes enough.
-->
Figured I'd try out the suggestion to perform intermediate reviews in my fork, @Zorro666. This is the same as the previous change you looked at but with a minor fix to the pipeline's depth format being inside a branch it shouldn't have been in. And also clang-format having been applied. 
I believe however that now that multiple rendertargets are bound to the shader-out and post-mod passes, this introduces a potential error due to the side effects by passes being written to the capture sources. I think this can be fixed by adjusting the RenderTargetWriteMask on these two passes.

jristic commented:
It seems I'm not understanding something about the shader out pass correctly because when I change the PSO setup to 
`
if(i == colorOutputIndex)
     pipeDesc.BlendState.RenderTarget[i].RenderTargetWriteMask = D3D12_COLOR_WRITE_ENABLE_ALL;
else
     pipeDesc.BlendState.RenderTarget[i].RenderTargetWriteMask = 0;
`
it causes the results to be wrong...
I guess another option would be to change the bound RT descriptors to NULL resources to prevent the writes from happening, assuming that doesn't cause a validation error.

Zorro666 commented:
Really nice work : before making PR to renderdoc I think we should do the following (I have tried this out locally and it is working):
- remove `GetPixelHistoryRenderTargetIndex()` which will cascade to also removing `uint32_t outputIndex` parameter from `GetPixelOcclusionPipeline()`, `ReplayDrawWithTests()` and `CreatePipeline()`. Along with TODOs about `outputIndex`

- Change this in `CreatePerFragmentPipeline()` when creating `pipes.shaderOutPipe`
```
for(uint32_t i = 0; i < D3D12_SIMULTANEOUS_RENDER_TARGET_COUNT; ++i)
{
  pipeDesc.BlendState.RenderTarget[i].BlendEnable = FALSE;
  if(i == colorOutputIndex)
    pipeDesc.BlendState.RenderTarget[i].RenderTargetWriteMask = D3D12_COLOR_WRITE_ENABLE_ALL;
  else
    pipeDesc.BlendState.RenderTarget[i].RenderTargetWriteMask = 0;
}
```
and then need to set the correct RenderTargetWriteMask setup for the `primitiveIdPipe` i.e.

```
// PrimitiveID pipeline only writes to RT0
for(uint32_t i = 1; i < D3D12_SIMULTANEOUS_RENDER_TARGET_COUNT; ++i)
  pipeDesc.BlendState.RenderTarget[i].RenderTargetWriteMask = 0;

pipeDesc.BlendState.RenderTarget[0].RenderTargetWriteMask = D3D12_COLOR_WRITE_ENABLE_ALL;
```

The restore is the piece which I think you might not have had when you tried to restrict the RenderTargetMask

Zorro666 commented:
I see these validation warnings (not errors) which we can have a think about in the future.
As long as the copy pixel is not copying from the unwanted channels / the unwanted channels are initialised to 0 that is fine or perhaps a write mask could be applied based on the number of active channels in the format?

```
D3D12 WARNING: ID3D12Device::CreateGraphicsPipelineState: Pixel Shader output 'SV_Target1' is writing to 1 components, while the corresponding RenderTarget slot [1] has format (R32G32B32A32_FLOAT) with 4 component(s). This results in undefined contents in the unwritten channels of the render target. [ STATE_CREATION WARNING #974: CREATEGRAPHICSPIPELINESTATE_PS_OUTPUT_RT_OUTPUT_MISMATCH]
D3D12 WARNING: ID3D12Device::CreateGraphicsPipelineState: Pixel Shader output 'SV_Target1' is writing to 1 components, while the corresponding RenderTarget slot [1] has format (R32G32B32A32_FLOAT) with 4 component(s). This results in undefined contents in the unwritten channels of the render target. [ STATE_CREATION WARNING #974: CREATEGRAPHICSPIPELINESTATE_PS_OUTPUT_RT_OUTPUT_MISMATCH]
```

Zorro666 commented:
Here is my local patch with the changes I suggested

[jake.patch](https://github.com/jristic/renderdoc/files/13901228/jake.patch)

jristic commented:
Thanks Jake, but I applied your patch and have the same problem:
![image](https://github.com/jristic/renderdoc/assets/1256627/3792394d-b065-486c-94b0-39f456518006)
It seems for anything other than RT0 the shader out for each primitive is wrong.

jristic commented:
Okay, it took me quite a bit of noodling around in my own samples but reproducing the problem outside of RenderDoc reminded me that IndependentBlendEnable needs to be set to True in order for the render target blend descs beyond the first to be used at all 😅
I'm working on a fix now.

jristic commented:
I've added the fix for RenderTargetWriteMask, and also changed the fixed color shader to only output to a single channel to fix these warnings:

> D3D12 WARNING: ID3D12Device::CreateGraphicsPipelineState: The Pixel Shader expects a Render Target View bound to slot 3, but the PSO indicates that none will be bound. This is OK, as writes of an unbound Render Target View are discarded. It is also possible the developer knows the data will not be used anyway. This is only a problem if the developer actually intended to bind a Render Target View here. [ STATE_CREATION WARNING 679: CREATEGRAPHICSPIPELINESTATE_RENDERTARGETVIEW_NOT_SET]

The only remaining warning is the one you mention, where the RT format and shader output don't match. It seems worth fixing to me if possible, if not for correctness then for validation cleanliness. 
I think it's also worth looking further into GetPixelOcclusionPipeline having the unused outputIndex. I don't see a warning resulting from it but there could be a bug hiding there.

jristic commented:
I've removed the unused parameter in GetPixelOcclusionPipeline as requested. Let me know if there's anything else to address before putting together a PR from these changes.

jristic commented:
Also I've updated the sample I'm using to test, as the previous version had a validation warning in it. It can be found here:
https://drive.google.com/file/d/1nJVSdIpgFpCvaWeAoiO6Tpa39UlRjdji/view?usp=sharing

Zorro666 commented on renderdoc/data/hlsl/d3d12_pixelhistory.hlsl:
```diff
@@ -150,22 +150,33 @@ float4 RENDERDOC_PrimitiveIDPS(uint prim : SV_PrimitiveID) : SV_Target0
 
 struct MultipleOutput
```
Very minor nitpick : this is not multiple output anymore and could be renamed but also it is a useful name to remind us that the output can go to different render targets but a single RT at a time

Zorro666 commented on renderdoc/driver/d3d12/d3d12_debug.cpp:
```diff
@@ -2916,10 +2916,14 @@ void D3D12Replay::PixelHistory::Init(WrappedID3D12Device *device, D3D12DebugMana
-                             D3DCOMPILE_WARNINGS_ARE_ERRORS, {}, "ps_5_0", &FixedColorPS);
-  shaderCache->GetShaderBlob(hlsl.c_str(), "RENDERDOC_PixelHistoryFixedColPS",
-                             D3DCOMPILE_WARNINGS_ARE_ERRORS, {}, "ps_6_0", &FixedColorPSDxil);
+  for(int i = 0; i < 8; ++i)
```
Style wise rather than use 8 : be good to use something that guarantees to match the size of the output arrays i.e. `ARRAY_COUNT(FixedColorPS)` or could use `D3D12_SIMULTANEOUS_RENDER_TARGET_COUNT`

Zorro666 commented on renderdoc/driver/d3d12/d3d12_debug.cpp:
```diff
@@ -2928,8 +2932,11 @@ void D3D12Replay::PixelHistory::Release()
   SAFE_RELEASE(PrimitiveIDPSDxil);
-  SAFE_RELEASE(FixedColorPS);
-  SAFE_RELEASE(FixedColorPSDxil);
+  for(int i = 0; i < 8; ++i)
```
Nice to replace 8 with the size of the array i.e. `ARRAY_COUNT(....)`

Zorro666 commented on renderdoc/driver/d3d12/d3d12_pixelhistory.cpp:
```diff
@@ -426,20 +426,24 @@ void D3D12DebugManager::PixelHistoryCopyPixel(ID3D12GraphicsCommandListX *cmd,
+                               ID3DBlob *FixedColorPSDxil[8])
+      : m_pDevice(device), m_PrimIDPS(PersistentPrimIDPS), m_PrimIDPSDxil(PersistentPrimIDPSDxil)
   {
+    for(int i = 0; i < 8; ++i)
```
Be good to replace the 8 with the array size

Zorro666 commented on renderdoc/driver/d3d12/d3d12_pixelhistory.cpp:
```diff
@@ -451,8 +455,8 @@ struct D3D12PixelHistoryShaderCache
   ID3DBlob *m_PrimIDPSDxil;
-  ID3DBlob *m_FixedColorPS;
-  ID3DBlob *m_FixedColorPSDxil;
+  ID3DBlob *m_FixedColorPS[8];
```
I think the 8 should be `D3D12_SIMULTANEOUS_RENDER_TARGET_COUNT`

Zorro666 reviewed (commented):
Thanks this is amazing work - great to get MRT working so quickly.
Suggested some minor style comments and it would be good to squash the history to remove any code formatting commits to make them part of the origin commits before making a PR to RenderDoc repo.

BTW I discovered you can upload files to PRs when you add a comment

Zorro666 commented on renderdoc/driver/d3d12/d3d12_replay.h:
```diff
@@ -503,8 +503,8 @@ class D3D12Replay : public IReplayDriver
 
-    ID3DBlob *FixedColorPS = NULL;
-    ID3DBlob *FixedColorPSDxil = NULL;
+    ID3DBlob *FixedColorPS[8] = {NULL};
```
Ideally the 8 would be `D3D12_SIMULTANEOUS_RENDER_TARGET_COUNT`

jristic commented:
Thanks for the feedback!

Yeah, I was planning to squash it all down to a single commit before going to PR. 

Will do with the array looping changes, though it's ironic that even Microsoft's headers declare their arrays this way:
`    D3D12_RENDER_TARGET_BLEND_DESC RenderTarget[ 8 ];`

Zorro666 commented:
Thank you for all the changes and yeah Microsoft code can be inconsistent I find about using hard coded values like 8 or actual constants/#define's.

I would say RenderDoc tries to use constants but it also has 8 in it quite a bit.

The rationale for my request to switch to the constant was because the constant was already being used in the pixel history code and it was doing loops using `D3D12_SIMULTANEOUS_RENDER_TARGET_COUNT` and not just 8.

I kind of precedence was set.
